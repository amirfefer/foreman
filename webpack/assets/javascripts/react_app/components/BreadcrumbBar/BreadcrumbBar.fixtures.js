export const breadcrumbItems = {
  items: [
    {
      caption: 'root',
      url: '/some-url',
    },
    {
      caption: 'child',
    },
  ],
};

export const breadcrumbTitleItems = {
  items: [
    {
      caption: 'title',
    },
  ],
};

export const resource = {
  resourceUrl: 'some/url',
  nameField: 'name',
  switcherItemUrl: 'some/url/:id',
};

export const resourceList = [
  { name: 'Host 1', url: '#' },
  { name: 'Host 2', url: '#' },
  { name: 'Host 3 with a very long name', url: '#' },
  { name: 'Host 4', url: undefined, onClick: jest.fn() },
  { name: 'Host 5', url: '#', onClick: undefined },
];

export const serverResourceListResponse = {
  data: {
    page: 1,
    total: 3,
    per_page: 2,
    results: [
      { id: '1', name: 'name-1' },
      { id: '2', name: 'name-2' },
      { id: '3', name: 'name-3' },
    ],
  },
};

export const breadcrumbSwitcherLoading = {
  loading: true,
  resources: [],
};

export const breadcrumbSwitcherLoaded = {
  loading: false,
  resources: resourceList,
};

export const breadcrumbSwitcherLoadedWithPagination = {
  ...breadcrumbSwitcherLoaded,
  currentPage: 2,
  totalPages: 3,
};

export const breadcrumbBar = {
  data: {
    resource,
    breadcrumbItems: breadcrumbItems.items,
    isSwitchable: false,
  },
};

export const breadcrumbBarSwithcable = {
  data: {
    resource,
    breadcrumbItems: breadcrumbItems.items,
    isSwitchable: true,
  },
};
