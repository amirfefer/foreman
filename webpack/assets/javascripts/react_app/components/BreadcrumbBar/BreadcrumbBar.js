import React from 'react';
import Breadcrumbs from './components/Breadcrumb';
import Switcher from './components/BreadcrumbSwitcher';
import API from '../../API';
import Helpers from '../../common/helpers';

class BreadcrumbBar extends React.Component {
  constructor(props) {
    super(props);
    this.state = { menu: [], resurceItems: [], switchable: true };
  }
  componentDidMount() {
    const { data: { reosurceUrl, name, menu, resource, action } } = this.props;
    API.get(reosurceUrl).then(({ data }) => {
      const newData = data.results.map(res => ({ name: res[name], url: Helpers.urlBuilder(resource, action, res.id) }));
      this.setState({ resurceItems: newData });
    });
    this.setState({ menu });
  }

  render() {
    const { menu, resurceItems } = this.state;
    return (
      <div>
        <Breadcrumbs data={{ menu }} >
        {this.state.switchable &&
        <Switcher resources={resurceItems} />
        }
        </Breadcrumbs>
        <hr className="breadcrumb-line" />
      </div>
    );
  }
}

export default BreadcrumbBar;
