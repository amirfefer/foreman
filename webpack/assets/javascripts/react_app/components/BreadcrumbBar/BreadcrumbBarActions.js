import API from '../../API';

import {
  BREADCRUMB_BAR_TOGGLE_SWITCHER,
  BREADCRUMB_BAR_CLOSE_SWITCHER,
  BREADCRUMB_BAR_RESOURCES_REQUEST,
  BREADCRUMB_BAR_RESOURCES_SUCCESS,
  BREADCRUMB_BAR_RESOURCES_FAILURE,
} from './BreadcrumbBarConstants';

export const toggleSwitcher = () => ({
  type: BREADCRUMB_BAR_TOGGLE_SWITCHER,
});

export const closeSwitcher = () => ({
  type: BREADCRUMB_BAR_CLOSE_SWITCHER,
});

export const loadSwitcherResourcesByResource = (resource, options = {}) => (dispatch) => {
  const {
    reosurceUrl, nameField, switcherItemUrl,
  } = resource;
  const { page = 1 } = options;

  const beforeRequest = () =>
    dispatch({
      type: BREADCRUMB_BAR_RESOURCES_REQUEST,
      payload: { resource, options },
    });

  const onRequestSuccess = response =>
    dispatch({ type: BREADCRUMB_BAR_RESOURCES_SUCCESS, payload: formatResults(response) });

  const onRequestFail = error =>
    dispatch({ type: BREADCRUMB_BAR_RESOURCES_FAILURE, payload: error });

  const formatResults = ({ data }) => {
    const switcherItems = Object.values(data.results).map((x, i) =>
      (data.results instanceof Array ?
        { name: x[nameField], url: switcherItemUrl.replace(':id', x.id) } :
        { name: x[i][nameField], url: switcherItemUrl.replace(':id', x[i].id) }));
    return {
      items: switcherItems,
      page: data.page,
      pages: Number(data.total) / Number(data.per_page),
    };
  };
  beforeRequest();

  return API.get(reosurceUrl, {}, { page }).then(onRequestSuccess, onRequestFail);
};
