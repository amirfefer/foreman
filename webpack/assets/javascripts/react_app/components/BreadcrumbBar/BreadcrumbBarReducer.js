import Immutable from 'seamless-immutable';

import {
  BREADCRUMB_BAR_TOGGLE_SWITCHER,
  BREADCRUMB_BAR_RESOURCES_REQUEST,
  BREADCRUMB_BAR_RESOURCES_SUCCESS,
  BREADCRUMB_BAR_RESOURCES_FAILURE,
} from './BreadcrumbBarConstants';

const initialState = Immutable({
  resourceSwitcherItems: [],
  isLoading: false,
  isSwitcherOpen: false,
  requestError: null,
});

export default (state = initialState, action) => {
  const { payload } = action;

  switch (action.type) {
    case BREADCRUMB_BAR_RESOURCES_REQUEST:
      return state
        .set('resourceSwitcherItems', [])
        .set('requestError', null)
        .set('isLoading', true);

    case BREADCRUMB_BAR_RESOURCES_SUCCESS:
      return state
        .set('resourceSwitcherItems', payload)
        .set('requestError', null)
        .set('isLoading', false);

    case BREADCRUMB_BAR_RESOURCES_FAILURE:
      return state
        .set('resourceSwitcherItems', [])
        .set('requestError', payload)
        .set('isLoading', false);

    case BREADCRUMB_BAR_TOGGLE_SWITCHER:
      return state.set('isSwitcherOpen', !state.isSwitcherOpen);

    default:
      return state;
  }
};
