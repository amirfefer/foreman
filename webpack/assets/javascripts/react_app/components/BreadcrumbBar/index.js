import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './BreadcrumbBarActions';
import reducer from './BreadcrumbBarReducer';

import BreadcrumbBar from './BreadcrumbBar';

// map state to props
const mapStateToProps = ({ breadcrumbBar }) => ({
  resourceSwitcherItems: breadcrumbBar.resourceSwitcherItems,
  isSwitcherOpen: breadcrumbBar.isSwitcherOpen,
  isLoadingResources: breadcrumbBar.isLoadingResources,
  requestError: breadcrumbBar.requestError,
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { breadcrumbBar: reducer };

// export connected component
export default connect(mapStateToProps, mapDispatchToProps)(BreadcrumbBar);
