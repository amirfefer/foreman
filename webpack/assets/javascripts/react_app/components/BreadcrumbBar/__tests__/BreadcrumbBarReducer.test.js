import {
  BREADCRUMB_BAR_TOGGLE_SWITCHER,
  BREADCRUMB_BAR_CLOSE_SWITCHER,
  BREADCRUMB_BAR_RESOURCES_REQUEST,
  BREADCRUMB_BAR_RESOURCES_SUCCESS,
  BREADCRUMB_BAR_RESOURCES_FAILURE,
} from '../BreadcrumbBarConstants';
import reducer from '../BreadcrumbBarReducer';

import { resource, resourceList } from '../BreadcrumbBar.fixtures';

const fixtures = {
  'should return the initial state': {},
  'should handle BREADCRUMB_BAR_TOGGLE_SWITCHER': {
    action: {
      type: BREADCRUMB_BAR_TOGGLE_SWITCHER,
    },
  },
  'should handle BREADCRUMB_BAR_CLOSE_SWITCHER': {
    action: {
      type: BREADCRUMB_BAR_CLOSE_SWITCHER,
    },
  },
  'should handle BREADCRUMB_BAR_RESOURCES_REQUEST': {
    action: {
      type: BREADCRUMB_BAR_RESOURCES_REQUEST,
      payload: {
        resource: { ...resource },
        options: {},
      },
    },
  },
  'should handle BREADCRUMB_BAR_RESOURCES_SUCCESS': {
    action: {
      type: BREADCRUMB_BAR_RESOURCES_SUCCESS,
      payload: [...resourceList],
    },
  },
  'should handle BREADCRUMB_BAR_RESOURCES_FAILURE': {
    action: {
      type: BREADCRUMB_BAR_RESOURCES_FAILURE,
      payload: new Error('some error'),
    },
  },
};

describe('BreadcrumbBar reducer', () => {
  const reduce = ({ state, action = {} } = {}) => reducer(state, action);

  Object.entries(fixtures).forEach(([description, action]) =>
    it(description, () => expect(reduce(action)).toMatchSnapshot()));
});
