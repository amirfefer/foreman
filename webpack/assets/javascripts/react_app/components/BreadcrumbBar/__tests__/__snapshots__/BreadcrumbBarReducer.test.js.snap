// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadcrumbBar reducer should handle BREADCRUMB_BAR_CLOSE_SWITCHER 1`] = `
Object {
  "currentPage": null,
  "isLoadingResources": false,
  "isSwitcherOpen": false,
  "pages": null,
  "requestError": null,
  "resourceSwitcherItems": Array [],
}
`;

exports[`BreadcrumbBar reducer should handle BREADCRUMB_BAR_RESOURCES_FAILURE 1`] = `
Object {
  "currentPage": null,
  "isLoadingResources": false,
  "isSwitcherOpen": false,
  "pages": null,
  "requestError": Object {},
  "resourceSwitcherItems": Array [],
}
`;

exports[`BreadcrumbBar reducer should handle BREADCRUMB_BAR_RESOURCES_REQUEST 1`] = `
Object {
  "currentPage": null,
  "isLoadingResources": true,
  "isSwitcherOpen": false,
  "pages": null,
  "requestError": null,
  "resourceSwitcherItems": Array [],
}
`;

exports[`BreadcrumbBar reducer should handle BREADCRUMB_BAR_RESOURCES_SUCCESS 1`] = `
Object {
  "currentPage": 1,
  "isLoadingResources": false,
  "isSwitcherOpen": false,
  "pages": 2,
  "requestError": null,
  "resourceSwitcherItems": Array [
    Object {
      "name": "Host 1",
      "url": "#",
    },
    Object {
      "name": "Host 2",
      "url": "#",
    },
    Object {
      "name": "Host 3 with a very long name",
      "url": "#",
    },
    Object {
      "name": "Host 4",
      "onClick": [Function],
      "url": undefined,
    },
    Object {
      "name": "Host 5",
      "onClick": undefined,
      "url": "#",
    },
  ],
}
`;

exports[`BreadcrumbBar reducer should handle BREADCRUMB_BAR_TOGGLE_SWITCHER 1`] = `
Object {
  "currentPage": null,
  "isLoadingResources": false,
  "isSwitcherOpen": true,
  "pages": null,
  "requestError": null,
  "resourceSwitcherItems": Array [],
}
`;

exports[`BreadcrumbBar reducer should return the initial state 1`] = `
Object {
  "currentPage": null,
  "isLoadingResources": false,
  "isSwitcherOpen": false,
  "pages": null,
  "requestError": null,
  "resourceSwitcherItems": Array [],
}
`;
